crates/bevy_ecs/src/bundle.rs

```diff
@@ -238,9 +251,25 @@ unsafe impl<C: Component> Bundle for C {
     }
 }
 
+/// SAFETY:
+// - `Bundle::from_components` calls `func` exactly once for C, which is the exact value returned by `Bundle::component_ids`.
+unsafe impl<C: Component> BundleFromComponents for C {
+    unsafe fn from_components<T, F>(ctx: &mut T, func: &mut F) -> Self
+    where
+        // Ensure that the `OwningPtr` is used correctly
+        F: for<'a> FnMut(&'a mut T) -> OwningPtr<'a>,
+        Self: Sized,
+    {
+        let ptr = func(ctx);
+        // Safety: The id given in `component_ids` is for `Self`
+        unsafe { ptr.read() }
+    }
+}
+
 impl<C: Component> DynamicBundle for C {
+    type Effect = ();
```

Comment:
Why is this `()` here, but `NoBundleEffect` elsewhere?

Reply:
NoBundleEffect is a trait that is implemented for empty tuples and all tuples of empty tuples. It is a way to identify any bundle with no effect.

Reply:
Ah, I see! I should have read the docs more closely.